File: EasySports-backen/src/main/java/com/easysports/model/Match.java
package com.easysports.model;

import com.easysports.enums.Sport;
import com.easysports.enums.MatchType;
import com.easysports.enums.MatchStatus;
import jakarta.persistence.*;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import java.time.LocalDateTime;

/**
 * Entidad central que representa un encuentro o partido.
 * Es el corazón de la lógica de negocio de la aplicación.
 */
@Entity
@Table(name = "encuentros")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Match {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /**
     * Código único y fácil de compartir para que otros jugadores se unan.
     * Se genera automáticamente en la capa de servicio.
     */
    @Column(name = "codigo_encuentro", nullable = false, unique = true, length = 10)
    private String codigo;

    /**
     * El tipo de deporte que se jugará en este encuentro.
     */
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Sport deporte;

    /**
     * Define si el partido es una reunión casual o un reto formal entre equipos.
     */
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private MatchType tipo;

    /**
     * El estado actual del ciclo de vida del encuentro.
     */
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private MatchStatus estado;

    /**
     * Fecha y hora programadas para el inicio del encuentro.
     */
    @Column(name = "fecha_programada", nullable = false)
    private LocalDateTime fechaProgramada;

    /**
     * El usuario que creó el encuentro.
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "creador_id", nullable = false)
    private User creador;

    /**
     * El equipo local. Es nulo si el partido es de tipo CASUAL.
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "equipo_local_id")
    private Team equipoLocal;

    /**
     * El equipo visitante. Es nulo si el partido es de tipo CASUAL.
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "equipo_visitante_id")
    private Team equipoVisitante;
    
    /**
     * La cancha donde se jugará el encuentro.
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "cancha_id")
    private Court cancha;

    /**
     * Número máximo de jugadores permitidos. Crucial para partidos casuales.
     */
    @Column(name = "max_jugadores", nullable = false)
    private Integer maxJugadores;

    /**
     * Contador de los jugadores actualmente inscritos.
     */
    @Column(name = "jugadores_actuales", nullable = false)
    private Integer jugadoresActuales = 0;
}

File: EasySports-flutter/easy_sports_app/lib/src/models/match.dart
class Match {
  final int id;
  final String codigo;
  final String tipo;
  final String deporte;
  final String estado;
  final DateTime fechaProgramada;
  final int? canchaId;
  final String? nombreCanchaTexto;
  final int creadorId;
  final int? equipoLocalId;
  final String? equipoLocalNombre;
  final int? equipoVisitanteId;
  final String? equipoVisitanteNombre;
  final int maxJugadores;
  final int jugadoresActuales;
  final int? golesLocal;
  final int? golesVisitante;
  final String? comentarios;

  Match({
    required this.id,
    required this.codigo,
    required this.tipo,
    required this.deporte,
    required this.estado,
    required this.fechaProgramada,
    this.canchaId,
    this.nombreCanchaTexto,
    required this.creadorId,
    this.equipoLocalId,
    this.equipoLocalNombre,
    this.equipoVisitanteId,
    this.equipoVisitanteNombre,
    required this.maxJugadores,
    required this.jugadoresActuales,
    this.golesLocal,
    this.golesVisitante,
    this.comentarios,
  });

  factory Match.fromJson(Map<String, dynamic> json) {
    return Match(
      id: json['id'],
      codigo: json['codigo'],
      tipo: json['tipo'],
      deporte: json['deporte'],
      estado: json['estado'],
      fechaProgramada: DateTime.parse(json['fechaProgramada']),
      canchaId: json['canchaId'],
      nombreCanchaTexto: json['nombreCanchaTexto'],
      creadorId: json['creadorId'],
      equipoLocalId: json['equipoLocalId'],
      equipoLocalNombre: json['equipoLocalNombre'],
      equipoVisitanteId: json['equipoVisitanteId'],
      equipoVisitanteNombre: json['equipoVisitanteNombre'],
      maxJugadores: json['maxJugadores'],
      jugadoresActuales: json['jugadoresActuales'],
      golesLocal: json['golesLocal'],
      golesVisitante: json['golesVisitante'],
      comentarios: json['comentarios'],
    );
  }
}
